sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c

```diff
@@ -0,0 +1,225 @@
+  // 0,1,2,3.
+  const uint32_t kExpectedOutput[kExpectedOutputLen] = {
+      0xe48bb8cb, 0x1012c84c, 0x5af8a7f1, 0xd1c07cd9, 0xdf82ab22, 0x771c619b,
+      0xd40fccb1, 0x87189e99, 0x510494b3, 0x64f7ac0c, 0x2581f391, 0x80b1dc2f,
+      0x793e01c5, 0x87b107ae, 0xdb17514c, 0xa43c41b7,
+  };
+
+  CHECK_ARRAYS_EQ(got, kExpectedOutput, kExpectedOutputLen,
+                  "Generate command KAT output mismatch");
+}
+
+/**
+ * Check that the internal states will read zero.
+ */
+static void test_fuse_disable(const dif_csrng_t *csrng) {
+  LOG_INFO("%s", __func__);
+  fips_instantiate_kat(csrng, /*fail_expected=*/true);
+}
+
+/**
+ * Read the otp at `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` address
+ * and check whether is configured by the `uvm_test_seq` as expected.
+ *
+ * @param expected Define the expected value for the
+ * HW_CFG.EN_ENTROPY_SRC_FW_READ flag.
+ */
+static void check_csrng_fuse_enabled(bool expected) {
+  dif_otp_ctrl_t otp;
+  CHECK_DIF_OK(dif_otp_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR), &otp));
+
+  dif_otp_ctrl_config_t config = {
+      .check_timeout = 100000,
+      .integrity_period_mask = 0x3ffff,
+      .consistency_period_mask = 0x3ffffff,
+  };
+  CHECK_DIF_OK(dif_otp_ctrl_configure(&otp, config));
+  otp_ctrl_testutils_wait_for_dai(&otp);
+
+  uint32_t value;
+  // Read the current value of the partition.
+  CHECK_DIF_OK(dif_otp_ctrl_dai_read_start(&otp, kDifOtpCtrlPartitionHwCfg,
+                                           kOtpIfetchHwRelativeOffset));
+  otp_ctrl_testutils_wait_for_dai(&otp);
+  CHECK_DIF_OK(dif_otp_ctrl_dai_read32_end(&otp, &value));
+  multi_bit_bool_t enable = bitfield_field32_read(
+      value,
+      (bitfield_field32_t){.mask = 0xff, .index = kOtpCsrngFwReadBitOffset});
+  CHECK((enable == kMultiBitBool8True) == expected,
+        "`fw_enable` not expected (%x)", enable);
+}
+
+/**
+ * This test:
+ *
+ * - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET`
+ * fuse bit set to enabled in the `uvm_test_seq`.
+ * - Issue an instantiate command to request entropy.
+ * - Verify that SW can read the internal states.
+ * - Reset the chip and repeat the steps above, but this time, with
+ *   `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit set to 0.
+ * - Verify that the SW reads back all zeros when reading the internal states.
+ */
+bool test_main(void) {
+  dif_csrng_t csrng;
+  CHECK_DIF_OK(dif_csrng_init(
+      mmio_region_from_addr(TOP_EARLGREY_CSRNG_BASE_ADDR), &csrng));
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+
+  dif_rstmgr_t rstmgr;
+  dif_rstmgr_reset_info_bitfield_t info;
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+  info = rstmgr_testutils_reason_get();
+
+  if (info == kDifRstmgrResetInfoPor) {
+    LOG_INFO("Powered up for the first time");
+    check_csrng_fuse_enabled(true);
```

Comment:
You check whether the fuse is true here, but where did you set it to begin with?

Reply:
If the default is true, then how is it unset for the else case below?

Reply:
The OTP can not be set by SW, so it is  done in the dv seq file (hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv).
This is documented before the main function.